File: packages/stage-director/src/stage_director/command_queue.py
#  Copyright (c) 2025 LongbowXXX
#
#  This software is released under the MIT License.
#  http://opensource.org/licenses/mit-license.php

import asyncio

from stage_director.models import StageCommand

# アプリケーション全体で共有するコマンドキュー
# StageCommand 型のオブジェクトを格納する
command_queue: asyncio.Queue[StageCommand] = asyncio.Queue()


# Add a dictionary to track events for specific commands
command_events: dict[str, asyncio.Event] = {}


async def enqueue_command(command: StageCommand) -> None:
    """MCPツールなどからコマンドをキューに入れる関数"""
    await command_queue.put(command)


async def dequeue_command() -> StageCommand:
    """WebSocketハンドラなどがコマンドをキューから取り出す関数"""
    return await command_queue.get()


def mark_command_done() -> None:
    """キュー内のタスク完了をマークする"""
    command_queue.task_done()


async def wait_for_command(command_id: str) -> None:
    """Wait for a specific command to be marked as done."""
    if command_id not in command_events:
        command_events[command_id] = asyncio.Event()
    await command_events[command_id].wait()


def notify_command_done(command_id: str) -> None:
    """Notify that a specific command is done."""
    if command_id in command_events:
        command_events[command_id].set()
        del command_events[command_id]

File: packages/stage-director/src/commandQueue.ts
//  Copyright (c) 2025 LongbowXXX
//
//  This software is released under the MIT License.
//  http://opensource.org/licenses/mit-license.php

import { StageCommand } from './models.js';

// Application-wide shared command queue
class CommandQueue {
  private queue: StageCommand[] = [];
  private waiters: Array<() => void> = [];
  private commandEvents: Map<string, () => void> = new Map();

  async enqueueCommand(command: StageCommand): Promise<void> {
    this.queue.push(command);
    if (this.waiters.length > 0) {
      const resolve = this.waiters.shift();
      if (resolve) {
        resolve();
      }
    }
  }

  async dequeueCommand(): Promise<StageCommand> {
    if (this.queue.length > 0) {
      const command = this.queue.shift();
      if (command) {
        return command;
      }
    }

    return new Promise((resolve) => {
      this.waiters.push(() => {
        const command = this.queue.shift();
        if (command) {
          resolve(command);
        }
      });
    });
  }

  async waitForCommand(commandId: string): Promise<void> {
    return new Promise((resolve) => {
      this.commandEvents.set(commandId, resolve);
    });
  }

  notifyCommandDone(commandId: string): void {
    const resolve = this.commandEvents.get(commandId);
    if (resolve) {
      resolve();
      this.commandEvents.delete(commandId);
    }
  }
}

export const commandQueue = new CommandQueue();
